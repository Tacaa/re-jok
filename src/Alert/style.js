import styled from 'styled-components';
import Icon from '../Icon';
import { hexa } from '../globals';

export const StyledIconClose = styled(Icon)`
 text-align: center;
 position: absolute;
 cursor: pointer;
 height: 1.6rem;
 width: 1.6rem;
 line-height: 1.6rem;
 vertical-align: middle;
 top: 1.6rem;
 right: 1.6rem;



 transition: all .2s;

`

export const StyledAlertIcon = styled(Icon)`
  margin-right: 10px;
  height: 1.6rem;
  width: 1.6rem;
  line-height: 1.6rem;
  vertical-align: middle;
`

const returnType = props => {
  if(!props.color){
    switch (props.type) {
      case 'none':
        return `
          color: ${props.theme.text.regular};
          background: ${ ({theme}) => theme.background.default};

        `
      case 'success':
        return`
          color: ${props.theme.success.main};
          background: ${hexa(props.theme.success.main, .28)};

        `
      case 'info':
        return`
          color: ${props.theme.info.main};
          background: ${hexa(props.theme.info.main, .28)};

        `
      case 'warning':
        return`
          color: ${props.theme.warning.main};
          background: ${hexa(props.theme.warning.main, .28)};

        `
      case 'error':
        return`
          color: ${props.theme.error.main};
          background: ${hexa(props.theme.error.main, .28)};

        `
      default:
        return `
          color: ${props.theme.text.regular};
          background: ${ ({theme}) => theme.background.default};

         `

    }
  }
  else {
    return `
    color: ${props.theme.text.default};
    background: ${props.color};
    `
  }

}

export const StyledAlert = styled.div`
  font-size: 1.4rem;
  position: relative;
  display: flex;
  align-items: center;
  padding: 1.5rem 3rem 1.5rem 1.5rem;
  margin-bottom: 20px;
  ${props => returnType(props)};
`

export const StyledAlertTitle = styled.span`
  font-size: 1.6rem;
  color: ${ ({theme}) => theme.text.primary};
`

export const StyledAlertContent =  styled.div`
  font-size: 1.4rem;
`
