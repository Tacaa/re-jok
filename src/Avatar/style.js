import styled, {css} from 'styled-components';

export const baseAvatar = css `
   display: flex;
   flex: none;
   align-self: center;
   align-items: center;
   justify-content: center;
   list-style: none;
   text-transform: uppercase;
   box-sizing: border-box;
   margin-right: 0.8rem;
   background: ${props => props.imgSrc
  ? "transparent"
  : "#bdbdbd"} ;
   color: #fafafa;
   font-size: ${props => props.theme.size[props.size].fontSize};
   line-height: ${props => props.theme.size[props.size].lineHeight};
   width: ${props => props.theme.size[props.size].width};
   height: ${props => props.theme.size[props.size].height};
   padding: 0;

`

export const StyledAvatar = styled.div `
  ${baseAvatar}

  border-radius: ${props => props.shape === "circle" ? '50%' : '.4rem'};

`
export const StyledAvatarImage = styled.img.attrs({
  src: props => props.imgSrc
    ? props.imgSrc
    : "",
  alt: props => props.imgAlt
    ? props.imgAlt
    : ""
})`

  border-radius: inherit;
  width: 100%;
  height: 100%;
  object-fit: cover;
  vertical-align: middle;
  border-style: none;
`

//=============== Avatar Group ==================
export const StyledAvatarGroup = styled.ul `
  list-style: none;
  display: inline-flex;
  margin-right: .8rem;
  margin-left: .8rem;
  pointer-events: auto;
  order: 1;
`

export const StyledAvatarItem = styled.li `
  margin-left: -.8rem;
  width: ${props => props.theme.size[props.size].width};
  height: ${props => props.theme.size[props.size].height};
  pointer-events: auto;
  box-shadow: rgb(255, 255, 255) 0px 0px 0px 2px;
  position: relative;
  border-radius: 50%;
  transition: transform 0.2s ease-in-out;
`
