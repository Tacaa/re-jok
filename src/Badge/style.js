import styled from 'styled-components'


export const StyledBadge = styled.span`
    font-size: 14px;
    line-height: 1.5;
    color: rgba(0,0,0,.65);
    box-sizing: border-box;
    margin: 0;
    padding: 0;
    list-style: none;
    position: relative;
    display: inline-flex;
    flex-direction: column;
    line-height: 1;
    vertical-align: middle;


`

export const StyledSup = styled.sup`
  background: ${props=> props.backgroundColor ? props.backgroundColor : "#f5222d"};
  position: absolute;
  transform: translateX(-50%);
  top: -1rem;
  left: ${props => props.widthChildren ? `${props.widthChildren/10}rem`: 0}
  height: 20px;
  border-radius: 10px;
  min-width: 20px;
  color: #fff;
  line-height: 20px;
  text-align: center;
  padding: ${props => props.count < 10 ? '0 .6rem' : '0 .8rem'};
  font-size: 12px;
  font-weight: 400;
  z-index: 1;
  white-space: nowrap;

  transform-origin: -10% center;


`
