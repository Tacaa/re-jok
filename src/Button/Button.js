// @flow
import * as React from 'react'

import Icon from '../Icon';

import {
  Btn,
  OutlineBtn,
  Label,
  SpinnerButton,
  IconBtn
} from './style';

import { isChild } from '../helpers/typeUtils';

type Props = {
  /** Set button unable to click or not **/
  disabled?: boolean,
  /** Event when user clicked to button **/
  onClick?: Function,
  /** Set button to a link**/
  hrefString?: string,
  /** The size of button . The default size of button is default **/
  size: 'small' | 'default' | 'medium' | 'large',
  /** Set shape of button **/
  shape?: 'circle' | 'rounded',
  /** Button can show a loading**/
  loading?: boolean,
  /** Set icon in button**/
  icon?: string,
  /** Add more style to button **/
  style?: Object,
  /** Add more class to button **/
  className?: string,
  /** Primary color of button **/
  primary?: boolean,
  /** Secondary color of button **/
  secondary?: boolean,
  /** Set width of button to 100%**/
  fullWidth?: boolean,
  /** move icon to left or right label. The default is left **/
  iconPosition: 'left' | 'right',
  /** if you don't want use primary or secondary , you can set your color to button **/
  backgroundColor?: string,
  /** Make your button animated**/
  animated?: string,
  /** Change tag name of button ... span, div, a v..v **/
  tag?: string,
  /** Change type of button**/
  variant: 'outline' | 'inverted' | 'raised' | 'icon',
  /** create button with box shadow or not**/
  hasBoxShadow: boolean,
  /** Set type of button **/
  rule: 'button' | 'reset' | 'submit',
  /**Children element can be string, number, HTML node anything you want**/
  children?: any,
}

const defaultProps = {
  disabled: false,
  size: 'default',
  rule: 'button',
  loading: false,
  primary: false,
  secondary: false,
  fullWidth: false,
  iconPosition: 'left',
  backgroundColor: "",
  tag: "button",
  variant: 'raised',
  hrefString: "",
  hasBoxShadow: false
};

class Button extends React.Component<Props> {
  static defaultProps = defaultProps;

  render() {
    const {
      tag,
      hrefString,
      children,
      loading,
      shape,
      size,
      primary,
      secondary,
      icon,
      iconPosition,
      variant,

    } = this.props;

   const ListButton = {
     raised: Btn,
     outline: OutlineBtn,
     inverted: OutlineBtn,
     icon: IconBtn
   }
   const hasChild = !isChild(children);

    const ElementType = !hrefString ? ListButton[variant].withComponent(tag) : ListButton[variant].withComponent('a')

    return (
      <ElementType {...this.props}>
        {
          icon
            ? (
              loading
              ? (<SpinnerButton size={size} iconPosition={iconPosition} shape={shape} primary={primary} secondary={secondary}/>)
              : (<Icon iconPosition={iconPosition} name={icon}/>))
            : ('')
        }
        {loading && !icon && <SpinnerButton size={size} iconPosition={iconPosition} shape={shape} primary={primary} secondary={secondary}/>}
        {
          shape === "circle" || variant === "icon"
            ? (null)
            : (hasChild && <Label>
              {children}
            </Label>)
        }

      </ElementType>
    )
  }
}

export default Button
