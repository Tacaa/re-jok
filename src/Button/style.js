import styled, {css, keyframes} from 'styled-components';

import {Transition, hexa, shadeColor} from '../globals'

export const baseButton = css `
    display: flex;
    flex: none;
    align-self: center;
    align-items: center;
    justify-content: center;
    word-break: keep-all;
    cursor: pointer;
    position: relative;
    outline: none;
    border: none;
    text-decoration: none;
    text-align: center;

    font-size: ${props => props.theme.size[props.size].fontSize};

      div + span,
      span + span {
        margin: 0 .8rem;
      }




// =======================Button=====================


`
const classicAnimated = css `
&:hover {
    transform: translateY(-.3rem);
    box-shadow: 0 1rem 2rem rgba(0,0,0,.2);

    &::after {
        transform: scaleX(1.4) scaleY(1.6);
        opacity: 0;
    }
}


&:active {
    outline: none;
    transform: translateY(-.1rem);
    box-shadow: 0 .5rem 1rem rgba(0,0,0,.2);
}
`

const basicAnimated = css `
&:hover{
  border-color:${props => props.backgroundColor
        ? hexa(props.backgroundColor, 0.9)
        : props.primary
          ? hexa(props.theme.primary.main, 0.9)
          : props.secondary
            ? hexa(props.theme.secondary.main, 0.9)
            : "#fff"};
  box-shadow: ${props => props.backgroundColor
              ? (`0 14px 26px -12px ${hexa(props.backgroundColor, 0.42)}, 0 4px 23px 0 rgba(0,0,0,0.12), 0 8px 10px -5px ${hexa(props.backgroundColor, 0.2)}`)

              : props.primary
                ? (`0 14px 26px -12px ${hexa(props.theme.primary.main, 0.42)}, 0 4px 23px 0 rgba(0,0,0,0.12), 0 8px 10px -5px ${hexa(props.theme.primary.main, 0.2)}`)

                : props.secondary
                  ? (`0 14px 26px -12px ${hexa(props.theme.secondary.main, 0.42)}, 0 4px 23px 0 rgba(0,0,0,0.12), 0 8px 10px -5px ${hexa(props.theme.secondary.main, 0.2)}`)

                  : (`0 14px 26px -12px rgba(0, 0, 0, 0.42), 0 4px 23px 0 rgba(0,0,0,0.12), 0 8px 10px -5px rgba(0, 0, 0, 0.2)`)};

}

&:active{
  box-shadow: -.1rem .1rem .4rem rgba(0, 0, 0, 0.4);
}
`

export const Btn = styled.button.attrs({
  href: props => props.hrefString
    ? props.hrefString
    : null,
  type: props => props.tag === "button"
    ? props.rule
    : null,
  disabled: props => props.disabled || props.loading
})`
  ${baseButton}


  background-color: ${props => props.backgroundColor
  ? props.backgroundColor
  : props.primary
    ? props.theme.primary.main
    : props.secondary
      ? props.theme.secondary.main
      : "#fff"};

  color: ${props => (props.backgroundColor || props.primary || props.secondary)
        ? "#fff"
        : "#000000a6"};

    width: ${props => props.fullWidth && props.shape !== "circle"
          ? "100%"
          : "unset"};


  border: .1rem solid rgba(0,0,0,.15);
  border-color: ${props => props.backgroundColor
            ? props.backgroundColor
            : props.primary
              ? props.theme.primary.main
              : props.secondary
                ? props.theme.secondary.main
                : "#d9d9d9"};


  ${
props => !props.shape
  ? css `
    padding: ${props => props.theme.size[props.size].padding};
    border-radius: ${props => props.theme.size[props.size].borderRadius};
    line-height: ${props => props.theme.size[props.size].lineHeight};

      `
  : props.shape === "circle" && props.variant !== 'icon'
    ? css `
      width: ${props => props.theme.size[props.size].width};
      height: ${props => props.theme.size[props.size].height};
      border-radius: 50%;
      padding: 0;
    `
    : css `
        border-radius: 10rem;
        padding: ${props => props.theme.size[props.size].padding};
        line-height: ${props => props.theme.size[props.size].lineHeight};
    `}

     transition: ${Transition.base.on};
${
  props => props.variant === 'icon'
  ? null
  : props.variant === "outline"
  ? null
  : (
    css`
    &:hover{
      background-color: ${props => props.backgroundColor
      ? hexa(props.backgroundColor, 0.9)
      : props.primary
        ? hexa(props.theme.primary.main, 0.9)
        : props.secondary
          ? hexa(props.theme.secondary.main, 0.9)
          : "#fff"};

    }
    `
    )
}

&::before {
position: absolute;
top: -1px;
left: -1px;
bottom: -1px;
right: -1px;
background: #fff;
opacity: .35;
content: "";
border-radius: inherit;
z-index: 1;
-webkit-transition: opacity .2s;
transition: opacity .2s;
pointer-events: none;
display: ${props => props.disabled || props.loading
      ? "block"
      : 'none'};
}

${props => props.variant === "raised"
        ? (
          !props.animated && props.hasBoxShadow
          ? basicAnimated
          : props.animated === "classic"
            ? classicAnimated
            : null)
        : null}

`

//==================== Outline Button ====================
export const OutlineBtn = Btn.extend `
  background-color: #fff;
  border-color: ${props => props.backgroundColor
  ? props.backgroundColor
  : props.primary
    ? props.theme.primary.main
    : props.secondary
      ? props.theme.secondary.main
      : "rgba(0,0,0,.16)"};
  color: ${props => props.backgroundColor
        ? props.backgroundColor
        : props.primary
          ? props.theme.primary.main
          : props.secondary
            ? props.theme.secondary.main
            : "#0000008a"};

  ${
props => props.variant === "outline"
  ? css `
       &:hover{
            border-color: ${props => props.backgroundColor
    ? shadeColor(props.backgroundColor, -15)
    : props.primary
      ? shadeColor(props.theme.primary.main, -15)
      : props.secondary
        ? shadeColor(props.theme.secondary.main, -15)
        : "rgba(0,0,0,.54)"};

            color: ${props => props.backgroundColor
          ? shadeColor(props.backgroundColor, -15)
          : props.primary
            ? shadeColor(props.theme.primary.main, -15)
            : props.secondary
              ? shadeColor(props.theme.secondary.main, -15)
              : "rgba(0,0,0,.68)"};
       }
     `
  : props.variant === "inverted"
    ? (
      css `
         &:hover{
           background-color: ${props => props.backgroundColor
      ? props.backgroundColor
      : props.primary
        ? props.theme.primary.main
        : props.secondary
          ? props.theme.secondary.main
          : "rgba(0,0,0,.54)"};
            color: #fff;
         }
         `)
    : null}

    ${ props => props.hasBoxShadow ?
      basicAnimated : null
    }



`

//=================== Icon Button =======================
export const IconBtn = Btn.extend `
   border: none;
   color: ${props => props.backgroundColor
  ? props.backgroundColor
  : props.primary
    ? props.theme.primary.main
    : props.secondary
      ? props.theme.secondary.main
      : "#0000008a"};
   width: ${props => props.theme.size[props.size].width};
   height: ${props => props.theme.size[props.size].height};
   border-radius: 50%;
   padding: 0;
   background-color: transparent;
   &:hover{
        border-color: ${props => props.backgroundColor
        ? shadeColor(props.backgroundColor, -15)
        : props.primary
          ? shadeColor(props.theme.primary.main, -15)
          : props.secondary
            ? shadeColor(props.theme.secondary.main, -15)
            : "rgba(0,0,0,.54)"};

        color: ${props => props.backgroundColor
              ? shadeColor(props.backgroundColor, -15)
              : props.primary
                ? shadeColor(props.theme.primary.main, -15)
                : props.secondary
                  ? shadeColor(props.theme.secondary.main, -15)
                  : "rgba(0,0,0,.68)"};
   }

`

//==================== Text Button =======================

const animationTextButton = css`
  &:before{
    content: "";
position: absolute;
bottom: 0;
left: 0;
right: 100%;
border-bottom: 1px solid #222;
transition: right .3s ease-out;
  }
  &:hover:before{
    right: 0;
  }
`


export const TextBtn = styled.button.attrs({
  href: props => props.hrefString
    ? props.hrefString
    : null,
  type: props => props.tag === "button"
    ? props.rule
    : null,
  disabled: props => props.disabled || props.loading
})`
 ${baseButton}
 letter-spacing: 0.08rem;
 background: transparent;
 background-image: none;

 color: ${props => props.color
    ? props.color
    : props.primary
      ? props.theme.primary.main
      : props.secondary
        ? props.theme.secondary.main
        : "#000000a6"};

 padding: ${props => props.theme.size[props.size].padding};
 &:hover{

   color: ${props => props.colorWhenHover ?
        hexa(props.colorWhenHover)
        :
         props.color
          ? hexa(props.color, 0.8)
          : props.primary
            ? hexa(props.theme.primary.main, 0.8)
            : props.secondary
              ? hexa(props.theme.secondary.main, 0.8)
              : hexa("#000000a6", 0.8)};

    border-color: transparent;

 }
 &:active{
   color: ${props => props.colorWhenHover ?
        hexa(props.colorWhenHover)
        : props.color
                  ? hexa(props.color, 0.8)
                  : props.primary
                    ? hexa(props.theme.primary.main, 0.8)
                    : props.secondary
                      ? hexa(props.theme.secondary.main, 0.8)
                      : hexa("#000000a6", 0.8)};
border-color: transparent;
 }

${
  props =>
   props.hoverable && animationTextButton
}



`


// =================== Sub Component for Button ========================

export const Label = styled.span `
  display: block;
  flex: 0 0 auto;
  line-height: inherit;
  color: inherit;
  align-self: center;
  margin: auto;
`;

const spin = keyframes `
  to {transform: rotate(360deg);}
`;

export const SpinnerButton = styled.div `
display: inline-block;
width: ${props => props.theme.size[props.size].fontSize};
height: ${props => props.theme.size[props.size].fontSize};
min-width: ${props => props.theme.size[props.size].fontSize};
min-height: ${props => props.theme.size[props.size].fontSize};
position: ${props => props.shape === "circle"
  ? ""
  : "relative"};
color: inherit;
flex: 0 0 ${props => props.theme.size[props.size].fontSize};
order: ${props => props.shape !== "circle"
    ? props.iconPosition === "left"
      ? 0
      : 1
    : 0};
  &::before{
    content: '';
  box-sizing: border-box;
  position: absolute;
  top: 50%;
  left: 50%;
  width: ${props => props.theme.size[props.size].fontSize};
  height: ${props => props.theme.size[props.size].fontSize};
  margin-top: calc(${props => props.theme.size[props.size].fontSize} / -2);
  margin-left: calc(${props => props.theme.size[props.size].fontSize} / -2);
  border-radius: 50%;
  border-color: ${props => props.primary || props.secondary
      ? "#fff"
      : "#d9d9d9"};
  border-top: 2px solid;
  border-right: 2px solid transparent;
  animation: ${spin} .6s linear infinite;
  }
`
