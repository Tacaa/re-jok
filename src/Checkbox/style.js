import styled, {css} from 'styled-components';


const baseCheckbox = css`
font-size: 14px;
line-height: 1.5;
color: ${({theme}) => theme.text.regular};
box-sizing: border-box;
padding: 0;
list-style: none;
display: inline-block;
position: relative;
white-space: nowrap;
cursor: pointer;
`

//Style Checkbox
const baseCheckboxButton = css`
height: 1.6rem;
width: 1.6rem;
border:  ${props => props.checked || props.indeterminate ? props.color ? `1px solid ${props.color}` : `1px solid ${props.theme.primary.main}` : "1px solid #e8e8e8"};
display: inline-block;
position: absolute;
left: 0;
background-color: ${({color, theme, checked, indeterminate}) => checked  || indeterminate ? color ? color : theme.primary.main : "#fff"};
top: 0;
${
  props => props.disabled && css`
  border-color: #d9d9d9!important;
  background-color: #f5f5f5;
  cursor: not-allowed;
  `
}

&::after{
  content: " ";
  position: absolute;
  display: table;

${
  props =>
  props.checked
  ?  css`
  left: 4.571429px;
  top: 1.142857px;
  width: 5.71428571px;
  height: 9.14285714px;
  transform: rotate(45deg) scale(1);

  `: props.indeterminate
  ? css`
  left: 2.42857143px;
  top: 5.92857143px;
  width: 9.14285714px;
  height: 1.14285714px;
  transform: scale(1);

  `: null
}

border: .2rem solid #fff;
border-top: 0;
border-left: 0;
opacity: ${props => (props.checked || props.indeterminate) && !props.disabled ? "1" : "0"};
transition: all .1s cubic-bezier(.71,-.46,.88,.6);

  ${
    props => props.disabled && css`

    background-color: #ccc;
    `
  }
}
`
//Style Toggle

const baseToggleButton = css`
    margin-right: .5rem;
    display: inline-block;
    position: relative;
    width: 4rem;
    height: 2rem;
    border: ${({checked, inactiveColor}) => !checked && inactiveColor ? `.1rem solid ${inactiveColor}` : ".1rem solid #dcdfe6"};
    outline: none;
    border-radius: 1rem;
    box-sizing: border-box;
    background: ${({checked, inactiveColor}) => !checked && inactiveColor ? inactiveColor : "#dcdfe6"};
    cursor: pointer;
    transition: border-color .3s,background-color .3s;
    vertical-align: middle;
    ${
      props => props.checked && !props.disabled && css`
      border-color: ${({color, theme, checked}) => checked ? color ? color : theme.primary.main : null};
      background-color: ${({color, theme, checked}) => checked ? color ? color : theme.primary.main : null};
      `
    }


    &::after{
      content: "";
      position: absolute;
      top: 1px;
      left: 1px;
      border-radius: 100%;
      transition: all .3s;
      width: 1.6rem;
      height: 1.6rem;
      background-color: #fff;
      ${
        props => props.checked && css`
          left: 100%;
          margin-left: -17px;
        `
      }
    }

`

export const StyledCheckboxGroup = styled.div`
  ${baseCheckbox};
  display: flex;
  flex-direction: ${props => props.layout === "horizontal" ? "row" : "column"};
`

export const StyledCheckboxWrapper = styled.div`
  ${baseCheckbox};
`


export const StyledCheckboxButton = styled.span`
  ${
    props =>
      props.toggle
      ? baseToggleButton
      : baseCheckboxButton
  }

`

export const StyledCheckbox = styled.input.attrs({
  type: "checkbox",

})`
position: fixed;
opacity: 0;
pointer-events: none;
margin: 0;
padding: 0;
width: 0;
height: 0;


`


export const StyledCheckboxLabel = styled.label.attrs({
  htmlFor: props=> props.id
})`
  font-size: 1.4rem;
  cursor: pointer;
  position: relative;
  padding-left: ${props => !props.toggle && "2rem"};
  padding-right: 2rem;
  color: ${props => props.disabled && "rgba(0,0,0,.25)"};
  cursor: ${props => props.disabled && "not-allowed"};

  &:hover ${StyledCheckboxButton}{
    ${
      props => !props.disabled && !props.toggle && css`
      border: ${({color, theme}) => color ? `1px solid ${color}` : `1px solid ${theme.primary.main}`};

      `
    }

  }
`
