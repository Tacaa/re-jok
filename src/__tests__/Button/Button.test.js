import React from 'react';
import Button from '../../Button/Button'
import {SpinnerButton} from '../../Button/style'
import {defaultTheme} from '../../theme'
import Icon from '../../Icon';
import { shallowWithTheme, mountWithTheme, renderWithTheme } from '../../helpers/withTheme';
import 'jest-enzyme'
import 'jest-styled-components'


describe('<Button />', () => {
  it('should render', () => {
    const tree = renderWithTheme(
      <Button>
        Button
      </Button>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a button correctly', () => {
    const wrapper = mountWithTheme(<Button/>, defaultTheme);
    expect(wrapper.find('button').type()).toEqual('button');
 });
 it('renders a button with text', () => {
   const wrapper = mountWithTheme(<Button>Button</Button>, defaultTheme);
   expect(wrapper.text()).toEqual("Button");
 });

 it('renders a button disabled', () => {
   const wrapper = mountWithTheme(<Button disabled={true}/>, defaultTheme);
   expect(wrapper.props().disabled).toEqual(true);
});

it('renders a button with defaultProps', () => {
  const wrapper = shallowWithTheme(<Button/>, defaultTheme);
  expect(wrapper.props().size).toEqual('default');
  expect(wrapper.props().loading).toEqual(false);
  expect(wrapper.props().rule).toEqual('button');
  expect(wrapper.props().variant).toEqual('raised');
  expect(wrapper.props().primary).toEqual(false);
  expect(wrapper.props().secondary).toEqual(false);
  expect(wrapper.props().fullWidth).toEqual(false);
  expect(wrapper.props().iconPosition).toEqual('left');


});

it('renders a button size small', () => {
  const wrapper = mountWithTheme(<Button size="small"/>, defaultTheme);
  expect(wrapper.props().size).toEqual('small');

  const tree = renderWithTheme(<Button size="small"/>).toJSON();
  expect(tree).toHaveStyleRule('font-size', '1.2rem')
  expect(tree).toHaveStyleRule('padding', '0.5rem 0.9rem')
  expect(tree).toHaveStyleRule('line-height', '1')
  expect(tree).toHaveStyleRule('border-radius', '.5rem')
});

it('renders a button size default', () => {
  const wrapper = mountWithTheme(<Button size="default"/>, defaultTheme);
  expect(wrapper.props().size).toEqual('default');

  const tree = renderWithTheme(<Button size="default"/>).toJSON();
  expect(tree).toHaveStyleRule('font-size', '1.6rem')
  expect(tree).toHaveStyleRule('padding', '0.8rem 1.2rem')
  expect(tree).toHaveStyleRule('line-height', '1.5')
  expect(tree).toHaveStyleRule('border-radius', '.8rem')
});

it('renders a button size medium', () => {
  const wrapper = mountWithTheme(<Button size="medium"/>, defaultTheme);
  expect(wrapper.props().size).toEqual('medium');

  const tree = renderWithTheme(<Button size="medium"/>).toJSON();
  expect(tree).toHaveStyleRule('font-size', '2rem')
  expect(tree).toHaveStyleRule('padding', '1.2rem 1.6rem')
  expect(tree).toHaveStyleRule('line-height', '1.5')
  expect(tree).toHaveStyleRule('border-radius', '.8rem')
});

it('renders a button size large', () => {
  const wrapper = mountWithTheme(<Button size="large"/>, defaultTheme);
  expect(wrapper.props().size).toEqual('large');

  const tree = renderWithTheme(<Button size="large"/>).toJSON();
  expect(tree).toHaveStyleRule('font-size', '2.4rem')
  expect(tree).toHaveStyleRule('padding', '1.4rem 1.8rem')
  expect(tree).toHaveStyleRule('line-height', '2')
  expect(tree).toHaveStyleRule('border-radius', '1rem')
});

it('renders a button loading true with spin component', () => {
  const wrapper = mountWithTheme(<Button loading={true}/>, defaultTheme);
  expect(wrapper.props().loading).toEqual(true);

  expect(wrapper.find(SpinnerButton)).toHaveLength(1);

});

it('renders a button loading false without spin component', () => {
  const wrapper = shallowWithTheme(<Button loading={false}/>, defaultTheme);
  expect(wrapper.props().loading).toEqual(false);

  expect(wrapper.find(SpinnerButton)).toHaveLength(0)

});

it('renders a button with icon', () => {
  const wrapper = shallowWithTheme(<Button icon="rocket"/>, defaultTheme);
  expect(wrapper.props().icon).toEqual("rocket");

  expect(wrapper.find(Icon)).toHaveLength(1)

});


it('should hide icon when button loading', () => {
  const wrapper = shallowWithTheme(<Button icon="rocket" loading={true}/>, defaultTheme);
  expect(wrapper.props().icon).toEqual("rocket");
  expect(wrapper.props().loading).toEqual(true);


  expect(wrapper.find(Icon)).toHaveLength(0)
  expect(wrapper.find(SpinnerButton)).toHaveLength(1)

});

it('renders a circle button and hide text of button', () => {
  const wrapper = mountWithTheme(<Button shape="circle">Button</Button>, defaultTheme);
  expect(wrapper.props().shape).toEqual("circle");
  expect(wrapper.text()).toEqual("");

  const tree = renderWithTheme(<Button shape="circle"/>).toJSON();
  expect(tree).toHaveStyleRule('width', '3.2rem')
  expect(tree).toHaveStyleRule('padding', '0')
  expect(tree).toHaveStyleRule('width', '3.2rem')
  expect(tree).toHaveStyleRule('border-radius', '50%')

});

it('renders a rounded button', () => {
  const wrapper = mountWithTheme(<Button shape="rounded"/>, defaultTheme);
  expect(wrapper.props().shape).toEqual('rounded');

  const tree = renderWithTheme(<Button shape="rounded"/>).toJSON();

  expect(tree).toHaveStyleRule('border-radius', '10rem')
});

it('renders a outline button', () => {
  const wrapper = mountWithTheme(<Button variant="outline"/>, defaultTheme);
  expect(wrapper.props().variant).toEqual('outline');

  const tree = renderWithTheme(<Button variant="outline"/>).toJSON();

  expect(tree).toHaveStyleRule('background-color', 'rgba(0,0,0,0)')
  expect(tree).toHaveStyleRule('border-color', 'rgba(0,0,0,.16)')
  expect(tree).toHaveStyleRule('color', '#0000008a')
  expect(tree).toHaveStyleRule('border-color', 'rgba(0,0,0,.54)',{
    modifier: ':hover',
  })
  expect(tree).toHaveStyleRule('color', 'rgba(0,0,0,.68)',{
    modifier: ':hover',
  })
});


it('renders a inverted button', () => {
  const wrapper = mountWithTheme(<Button variant="inverted"/>, defaultTheme);
  expect(wrapper.props().variant).toEqual('inverted');

  const tree = renderWithTheme(<Button variant="inverted"/>).toJSON();

  expect(tree).toHaveStyleRule('background-color', 'rgba(0,0,0,0)')
  expect(tree).toHaveStyleRule('border-color', 'rgba(0,0,0,.16)')
  expect(tree).toHaveStyleRule('color', '#0000008a')
  expect(tree).toHaveStyleRule('background-color', 'rgba(0,0,0,.54)',{
    modifier: ':hover',
  })
  expect(tree).toHaveStyleRule('color', '#fff',{
    modifier: ':hover',
  })
});


it('renders a icon button', () => {
  const wrapper = mountWithTheme(<Button variant="icon"/>, defaultTheme);
  expect(wrapper.props().variant).toEqual('icon');

  const tree = renderWithTheme(<Button variant="icon"/>).toJSON();

  expect(tree).toHaveStyleRule('border', 'none')
  expect(tree).toHaveStyleRule('border-radius', '50%')
  expect(tree).toHaveStyleRule('width', '3.2rem')
  expect(tree).toHaveStyleRule('height', '3.2rem')
  expect(tree).toHaveStyleRule('background-color', '#fff')
  expect(tree).toHaveStyleRule('border-color', 'rgba(0,0,0,.54)',{
    modifier: ':hover',
  })
  expect(tree).toHaveStyleRule('color', 'rgba(0,0,0,.68)',{
    modifier: ':hover',
  })
});


it('renders a link if pass props hrefString to button', () => {
  const wrapper = mountWithTheme(<Button hrefString="https://www.google.com/"/>, defaultTheme);
  expect(wrapper.props().hrefString).toEqual('https://www.google.com/');
  expect(wrapper.find('a').type()).toEqual('a');

});


it('renders a button fofllow by tag', ()=>{
  const wrapper = mountWithTheme(<Button tag="span">Button</Button>, defaultTheme);
  expect(wrapper.props().tag).toEqual('span');
  expect(wrapper.find('span')).toHaveLength(2);
})












});
