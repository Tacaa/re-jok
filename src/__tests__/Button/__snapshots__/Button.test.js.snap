// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render 1`] = `
.c1 {
  display: block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  line-height: inherit;
  color: inherit;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin: auto;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  word-break: keep-all;
  cursor: pointer;
  position: relative;
  outline: none;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  margin-right: 0.8rem;
  margin-bottom: 1.2rem;
  font-size: 1.6rem;
  background-color: #fff;
  color: #000000a6;
  width: unset;
  border: .1rem solid rgba(0,0,0,.15);
  border-color: #d9d9d9;
  padding: 0.8rem 1.2rem;
  border-radius: 1rem;
  line-height: 1.5;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
}

.c0 div + span,
.c0 span + span {
  margin: 0 .8rem;
}

.c0:hover {
  background: #fff;
}

.c0::before {
  position: absolute;
  top: -1px;
  left: -1px;
  bottom: -1px;
  right: -1px;
  background: #fff;
  opacity: .35;
  content: "";
  border-radius: inherit;
  z-index: 1;
  -webkit-transition: opacity .2s;
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
  pointer-events: none;
  display: none;
}

<button
  className="c0"
  disabled={false}
  href={null}
  size="default"
  type="button"
>
  
  <span
    className="c1"
  >
    Button
  </span>
</button>
`;
